### PR TITLE
Sort Lines As Integer Fix (issue#11023 )

### LINKED ISSUE #11023: Integers are not sorted correctly
### Description of the Issue

Negative numbers are sorted in reverse order

### Steps to Reproduce the Issue

1. Open attached file
[numbers.txt](https://github.com/notepad-plus-plus/notepad-plus-plus/files/7841344/numbers.txt)
2. Click 'Edit' in toolbar
3. Click 'Line Operations' in Edit menu
4. Click 'Sort Lines As Integers Descending ' or 'Sort Lines As Integers Ascending ' in Line Operations submenu

### Expected Behavior

Numbers are sorted ascending or descending (depending on option in step 4)

### Actual Behavior

Non negative numbers are sorted correctly while negative numbers are sorted in reverse order (ascending for descending sort and vice versa)

### Debug Information

Notepad++ v8.1.9.3   (64-bit)
Build time : Dec  6 2021 - 19:21:37
Path : C:\Program Files\Notepad++\notepad++.exe
Command Line : 
Admin mode : OFF
Local Conf mode : OFF
Cloud Config : OFF
OS Name : Windows 10 Pro (64-bit) 
OS Version : 2009
OS Build : 19042.1415
Current ANSI codepage : 1250
Plugins : mimeTools.dll NppConverter.dll NppExport.dll

### PR DESCRIPTION
**Pull request by sherwinnie**
Fix #11023. The current implementation is a variant of Natural Sort, which takes in account of negative numbers.

Some Examples:
```
-9223372036854775807
-9223372036854775808
s 99
s 100
-9223372036854775809
-9223372036854775808922337203685477580892233720368547758089223372036854775808
s -9223372036854775809
92233720368547758029223372036854775808 s
-922337203685477581092233720368547758029223372036854775808 s
-9223372036854775802
s -9223372036854775802
9223372036854775809 s
9223372036854775808 s
-9223372036854775804
-9223372036854775805
-92233720368547758039223372036854775808
s -9223372036854775810
9223372036854775807 s
s 101
92233720368547758029223372036854775808 s
```
is sorted ascendingly to:
```
-9223372036854775808922337203685477580892233720368547758089223372036854775808
-922337203685477581092233720368547758029223372036854775808 s
-92233720368547758039223372036854775808
-9223372036854775809
-9223372036854775808
-9223372036854775807
-9223372036854775805
-9223372036854775804
-9223372036854775802
9223372036854775807 s
9223372036854775808 s
9223372036854775809 s
92233720368547758029223372036854775808 s
92233720368547758029223372036854775808 s
s -9223372036854775810
s -9223372036854775809
s -9223372036854775802
s 99
s 100
s 101
```

Also, the following:
```
3
10
08
-03
06
5
004
003
-5
-009
010
-00
-7
-001
-004
-01
-02
-3
-005
04
1
07
03
006
01
-10
05
-007
0
-06
9
-010
-4
8
001
009
-09
-04
005
-0010
6
-07
-6
-0
-003
-002
-1
000
2
-006
-9
0010
-008
008
-000
-2
-8
09
7
-08
002
007
4
02
-05
00
```
is sorted to:
```
-0010
-010
-10
-009
-09
-9
-008
-08
-8
-007
-07
-7
-006
-06
-6
-005
-05
-5
-004
-04
-4
-003
-03
-3
-002
-02
-2
-001
-01
-1
-000
-00
-0
000
00
0
001
01
1
002
02
2
003
03
3
004
04
4
005
05
5
006
06
6
007
07
7
008
08
8
009
09
9
0010
010
10
```

I have tested other test cases, but they are way to long to be pasted here. Is there a place where I can show the test cases?

### CODE REVIEW

File: PowerEditor/src/MISC/Common/Sorters.h [hunk 1]
```diff
@@ -290,6 +290,193 @@ class NaturalSorter : public ISorter
...
+						bChunkIsNum = (b[bNumIndex] == L'-' && (b[bNumIndex + 1] >= L'0' && b[bNumIndex + 1] <= L'9'));
+						bNumSign = -1;
+					}
+
+					// One is number and one is string
+					if (aChunkIsNum != bChunkIsNum)
+					{
+						compareResult = a[aNumIndex] - b[bNumIndex];
+
+						// compareResult isn't nesscarily 0
```
Comment (alankilborn) [inline PowerEditor/src/MISC/Common/Sorters.h:341]: necessarily

File: PowerEditor/src/MISC/Common/Sorters.h [hunk 2]
```diff
@@ -505,7 +505,7 @@ class IntegerSorter : public ISorter
 					{
 						compareResult = a[aNumIndex] - b[bNumIndex];
 
-						// compareResult isn't nesscarily 0
+						// compareResult isn't necessarily 0
```
Comment (donho) [inline PowerEditor/src/MISC/Common/Sorters.h:508]: It's not into this PR - The `Files changed` view shown the typo still there.

Comment (sherwinnie) [reply PowerEditor/src/MISC/Common/Sorters.h:508]: Sorry, I forgot there is a copied and pasted code for ColumnSort and only fixed one typo. It should be fixed now.

File: PowerEditor/src/NppCommands.cpp [hunk 3]
```diff
@@ -708,7 +708,7 @@ void Notepad_plus::command(int id)
 			}
 			else if (id == IDM_EDIT_SORTLINES_INTEGER_DESCENDING || id == IDM_EDIT_SORTLINES_INTEGER_ASCENDING)
 			{
-				pSorter = std::unique_ptr<ISorter>(new NaturalSorter(isDescending, fromColumn, toColumn));
```
Comment (donho) [inline PowerEditor/src/NppCommands.cpp:711]: NaturalSorter class is not necessary anymore - it should be removed from the code.

**Review by donho — CHANGES_REQUESTED** [review]
Remove unnecessary codes.

File: PowerEditor/src/NppCommands.cpp (hunk 3 above)
Comment (sherwinnie) [reply PowerEditor/src/NppCommands.cpp:711]: Should I delete NaturalSorter from the code entirely? ie, delete its implementation from sorter.h? Or should I just delete its usage in NppCommands.cpp?

Comment (donho) [reply PowerEditor/src/NppCommands.cpp:711]: NaturalSorter should be removed from the code entirely, since it is not used anymore.

Comment (sherwinnie) [reply PowerEditor/src/NppCommands.cpp:711]: NaturalSorter is now removed from Sorter.h.

**Pull request closed without merging**